kernel/libs/page_allocator/src/buddy_allocator.rs

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::collections::btree_set::BTreeSet;
+use core::{alloc::Layout, array, cmp::min};
+
+use ostd::mm::{page::allocator::PageAlloc, Paddr, PAGE_SIZE};
+
+/// # Buddy Frame allocator
+///
+/// originated from crate `buddy_system_allocator`
+///
+/// ## Introduction
+///
+/// The max order of the allocator is specified via the const generic parameter
+/// `ORDER`. The frame allocator will only be able to allocate ranges of size
+/// up to 2<sup>ORDER</sup>, out of a total range of size at most 2<sup>ORDER +
+/// 1</sup> - 1.
+pub struct BuddyFrameAllocator<const ORDER: usize = 32> {
+    // buddy system with max order of ORDER
+    free_list: [BTreeSet<usize>; ORDER],
+
+    // statistics
+    pub allocated: usize,
+    pub total: usize,
+}
+
+pub(crate) fn prev_power_of_two(num: usize) -> usize {
+    1 << (usize::BITS as usize - num.leading_zeros() as usize - 1)
+}
+
+impl<const ORDER: usize> BuddyFrameAllocator<ORDER> {
+    /// Create an empty frame allocator
+    pub fn new() -> Self {
+        Self {
+            free_list: array::from_fn(|_| BTreeSet::default()),
+            allocated: 0,
+            total: 0,
+        }
+    }
+
+    /// Add a range of frame number [start, end) to the allocator
+    pub fn add_frame(&mut self, start: usize, end: usize) {
+        assert!(start <= end);
+
+        let mut total = 0;
+        let mut current_start = start;
+
+        while current_start < end {
+            let lowbit = if current_start > 0 {
+                current_start & (!current_start + 1)
+            } else {
+                32
+            };
+            let size = min(
+                min(lowbit, prev_power_of_two(end - current_start)),
+                1 << (ORDER - 1),
+            );
+            total += size;
+
+            self.free_list[size.trailing_zeros() as usize].insert(current_start);
+            current_start += size;
+        }
+
+        self.total += total;
+    }
+
+    /// Allocate a range of frames from the allocator, returning the first frame of the allocated
+    /// range.
+    pub fn alloc(&mut self, count: usize) -> Option<usize> {
```

Comment:
You should differentiate `count` and `size`, the former should mean the number of pages to be allocated (which is not quite good), and the latter should mean the total number of bytes of the pages to be allocated.